github.cake

```diff
@@ -0,0 +1,78 @@
+#addin "nuget:https://api.nuget.org/v3/index.json?package=Octokit&version=0.48.0"
+#addin "nuget:https://api.nuget.org/v3/index.json?package=NuGet.Versioning&version=5.7.0"
+#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.Yaml&version=3.1.1"
+
+using Octokit;
+using NuGet.Versioning;
+
+static CakeGitHubReleaseInfo _cakeGitHubReleaseInfo;
+
+public static CakeGitHubReleaseInfo GetCakeGitHubReleaseInfo(this ICakeContext context, DirectoryPath releaseDir)
+{
+    if (!(_cakeGitHubReleaseInfo is null))
+    {
+        return _cakeGitHubReleaseInfo;
+    }
+
+    // We cache the latest Cake release information in a file to speed up local development and also
+    // to decrease the chance of getting rate limited by GitHub when running unauthenticated
+    var cachedGitHubDir = context.MakeAbsolute(releaseDir.Combine("cake-github"));
+    var cachedReleaseInfoFile = cachedGitHubDir.CombineWithFilePath($"{nameof(CakeGitHubReleaseInfo)}.yml");
+    if (context.FileExists(cachedReleaseInfoFile))
+    {
+        context.Verbose("Retrieving Cake release information from cached file {0}...", cachedReleaseInfoFile);
+
+        _cakeGitHubReleaseInfo = context.DeserializeYamlFromFile<CakeGitHubReleaseInfo>(cachedReleaseInfoFile);
+        LogGitHubReleaseInfo(context, _cakeGitHubReleaseInfo);
+
+        return _cakeGitHubReleaseInfo;
+    }
+
+    context.Information("Retrieving Cake release information from GitHub...");
+
+    var client = new GitHubClient(new ProductHeaderValue("cake-build-website"));
+
+    var gitHubAccessToken =  context.EnvironmentVariable("git_access_token");
```

Comment:
You can use the global `accessToken` variable which is set in the `build.cake` file here

Reply:
My intent was to move everything GitHub-related to the `github.cake` file, thus I believe this variable no longer exists in `build.cake` - unless I forgot to delete, in which case I will.

Reply:
Indeed I forgot to delete the `accessToken` variable from `build.cake`. I deleted it just now and updated the PR. My thinking here is that the dependency is top-down i.e. `build.cake` depends on `github.cake` so it knows what to expect from it, whereas `github.cake` doesn't have any knowledge of who is including them, so expecting a global variable `accessToken` seems like an obscure dependency for `github.cake` to have.

That said, if the goal of the separate `.cake` files is _not_ to have reusable code across `.cake` files, but just make `build.cake` smaller, perhaps it would make sense to rename the other files to something like `build._nuget.cake` & `build._github.cake` to make the relationship more explicit (with the underscore hinting it's a private dependency).

Reply:
> My thinking here is that the dependency is top-down i.e. `build.cake` depends on `github.cake` so it knows what to expect from it, whereas `github.cake` doesn't have any knowledge of who is including them, so expecting a global variable `accessToken` seems like an obscure dependency for `github.cake` to have.
>
> That said, if the goal of the separate `.cake` files is _not_ to have reusable code across `.cake` files, but just make `build.cake` smaller, perhaps it would make sense to rename the other files to something like `build._nuget.cake` & `build._github.cake` to make the relationship more explicit (with the underscore hinting it's a private dependency).

You should not confuse Cake scripts with e.g. classes. It's really as simple as the scripts mentioned with `#load` are concatenated together in one big file. There's no separation or scoping going on here. A global variable, regardless in which script it is declared, will be available in any script since in the end it will be concatenated into one single file. So the reason fo splitting into multiple `.cake` files  is only to make files smaller and code better readable.  To decouple dependencies of the code, keeping state in a typed context would be the best way, since with this state can be passed across tasks and can also into helper methods. There won't be any global variables in this case and therefore also no more dependency.

But same as in the comment above. I'm OK with merging it like this and refactoring of build or introduction of a typed context can be done in a later PR.

Reply:
I understand the concat process, and the scoping. What I tried to convey is:

1. It's not very clear which scripts are meant to be executed directly (e.g. `build.cake`, `deploy.cake`) and which scripts are meant to be loaded by other scripts only (`nuget.cake` and now `github.cake`) so perhaps we could use different file names to make the intent (and maybe the relationship) more explicit

2. If one of those separate files (e.g. `github.cake`) would use a variable that is not declared within the file, it's not very clear where the variable is declared (could be `build.cake`, `nuget.cake`, ...) If `deploy.cake` needed some `github.cake` function, now it would also have to declare the same global variables that `github.cake` expects it to have, and declare it the same way

Of course, I'm looking at the above in the shoes of new contributors... I'm familiar enough now with the website build scripts to know where things are and what the execution chain is.

Reply:
1: The way I've solved this and seen in other repos is with a  `build.cake` file in the root folder and other Cake files loaded by the `build.cake` file in a `build` subfolder. 
2: The thing is it doesn't matter where global variables are declared. If a variable is declared in `build.cake` and used in `nuget.cake`, then it will also work if the method in `nuget.cake` is called from `deploy.cake`. I agree with the point that code can be hard to read with global variables. Writing helper methods in a way that everything is passed in as arguments would get rid of the dependencies. It's actually like with any .NET program :)